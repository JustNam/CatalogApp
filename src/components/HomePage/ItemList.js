/* eslint-disable arrow-parens */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Pagination } from 'react-bootstrap';
import CreateItemModal from '../Modals/CreateItemModal';
import SuccessModal from '../Modals/SuccessModal';
import { historyWithRefresh } from '../../history';

class ItemList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfoModal: false,
      showSuccessModal: false,
    };
  }

  componentDidMount() {
    this.props.getCategories();
  }

  handleInfoShow = () => {
    this.setState({ showInfoModal: true });
  };

  handleInfoClose = () => {
    this.setState({ showInfoModal: false });
  };

  handleSuccessShow = () => {
    this.setState({ showSuccessModal: true });
  };

  handleSuccessClose = () => {
    this.setState({ showSuccessModal: false });
    historyWithRefresh.push('/categories');
  };

  getNewPage = (page) => {
    const { item } = this.props;
    this.props.getItemsInCategoryWithPagination(item.categoryId, page);
  }

  createPagination = () => {
    const { item } = this.props;
    const paginations = [];
    [...Array(item.lastPage).keys()].forEach((index) => {
      const page = index + 1;
      if (page === item.currentPage) {
        paginations.push(
          <Pagination.Item id={page} key={page} active>
            {page}
          </Pagination.Item>
        );
      } else {
        paginations.push(
          <Pagination.Item
            id={page}
            onClick={() => this.getNewPage(page)}
            key={page}
          >
            {page}
          </Pagination.Item>
        );
      }
    });
    return <Pagination>{paginations}</Pagination>;
  }

  renderItems(currentCategory, item) {
    const { showInfoModal, showSuccessModal } = this.state;
    return (
      <div>
        <div>
          <div className="list-header">
            {currentCategory[0] && `category ${currentCategory[0].name}`}
            <div className="pull-xs-right">
              <button
                id="createButton"
                className="btn btn-sm btn-outline-primary"
                type="button"
                onClick={this.handleInfoShow}
              >
                Create
              </button>
            </div>
          </div>
        </div>
        {item.data.map((itemDetail) => (
          <div className="article-preview" key={itemDetail.id}>
            <Link
              className="preview-link"
              to={`/categories/${item.categoryId}/items/${itemDetail.id}`}
            >
              <h1>{itemDetail.title}</h1>
              {(itemDetail.description && (
              <p>{(itemDetail.description.length > 100) ? `${itemDetail.description.substring(0, 100)}...` : itemDetail.description}</p>)
              )}
              <span>Read more</span>
            </Link>
          </div>
        ))}
        {this.createPagination()}
        <CreateItemModal
          currentCategoryId={currentCategory[0].id}
          show={showInfoModal}
          handleClose={this.handleInfoClose}
          showSuccessModal={this.handleSuccessShow}
        />
        <SuccessModal
          show={showSuccessModal}
          handleClose={this.handleSuccessClose}
        />
      </div>
    );
  }

  render() {
    const { item, category } = this.props;
    const currentCategory = category.data.filter(
      (categoryDetail) => parseInt(item.categoryId) === parseInt(categoryDetail.id)
    );
    if (item.data.length !== 0) {
      return this.renderItems(currentCategory, item);
    }
    return (
      <div>
        <div className="list-header">
          {currentCategory[0] && `category ${currentCategory[0].name}`}
          <div className="pull-xs-right">
            <button
              id="createButton"
              className="btn btn-sm btn-outline-primary"
              type="button"
              onClick={this.handleInfoShow}
            >
              Create
            </button>
          </div>
        </div>
        <div className="article-preview">
          <a className="preview-link">
            <i>
              <span>
                This category is empty. Please select another category!
              </span>
            </i>
          </a>
        </div>
        {currentCategory[0]
        && (
        <CreateItemModal
          currentCategoryId={currentCategory[0].id}
          show={this.state.showInfoModal}
          handleClose={this.handleInfoClose}
          showSuccessModal={this.handleSuccessShow}
        />
        )
        }
      </div>
    );
  }
}

export default ItemList;
