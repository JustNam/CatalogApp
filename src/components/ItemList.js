/* eslint-disable arrow-parens */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import CreateItemModal from './modals/ItemInformationModal';
import SuccessModal from './modals/SuccessModal';

class ItemList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfoModal: false,
      showSuccessModal: false,
    };
  }

  handleInfoShow = () => {
    this.setState({ showInfoModal: true });
  };

  handleInfoClose = () => {
    this.setState({ showInfoModal: false });
  };

  handleSuccessShow = () => {
    this.setState({ showSuccessModal: true });
  };

  handleSuccessClose = () => {
    this.setState({ showSuccessModal: false });
  };

  renderItems(currentCategory, item) {
    const { showInfoModal, showSuccessModal } = this.state;
    return (
      <div>
        <div>
          <div className="list-header">
            {currentCategory[0] && `category ${currentCategory[0].name}`}
            <div className="pull-xs-right">
              <button
                className="btn btn-sm btn-outline-primary"
                type="button"
                onClick={this.handleInfoShow}
              >
                Create
              </button>
            </div>
          </div>
        </div>
        {item.data.map((itemDetail) => (
          <div className="article-preview" key={itemDetail.id}>
            <Link
              className="preview-link"
              to={`/categories/${item.categoryId}/items/${itemDetail.id}`}
            >
              <h1>{itemDetail.title}</h1>
              <p>{itemDetail.description}</p>
              <span>Read more...</span>
            </Link>
          </div>
        ))}
        <CreateItemModal
          currentCategoryId={currentCategory[0].id}
          show={showInfoModal}
          handleClose={this.handleInfoClose}
          showSuccessModal={this.handleSuccessShow}
        />
        <SuccessModal show={showSuccessModal} handleClose={this.handleSuccessClose} />
      </div>
    );
  }

  render() {
    const { item, category } = this.props;
    const currentCategory = category.data.filter(
      (categoryDetail) => item.categoryId === categoryDetail.id
    );
    if (item.data.length !== 0) {
      return this.renderItems(currentCategory, item);
    }
    return (
      <div>
        <p>{currentCategory[0] && `category ${currentCategory[0].name}`}</p>
        <div className="article-preview">
          <a className="preview-link" href="cafe">
            <i>
              <span>
                This category is empty. Please select another category!
              </span>
            </i>
          </a>
        </div>
      </div>
    );
  }
}

export default ItemList;
