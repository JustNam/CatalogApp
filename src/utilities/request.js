export const callAPI = async (enpoint) => {
  const response = await fetch(enpoint);
  const json = await response.json();
  return json;
};
