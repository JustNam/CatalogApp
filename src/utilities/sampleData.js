export const errorItem = {
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4)',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const itemWithEmptyCategory = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [],
};

export const noDescriptionItem = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const item = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 3,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 8,
      title: 'Ru3',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply',
      id: 9,
      title: 'Ru2',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const category = {
  data: [
    {
      created_on: '06/26/2019, 08:09:32',
      id: 1,
      name: 'Swimming',
      updated_on: '06/26/2019, 08:09:32',
    },
    {
      created_on: '06/26/2019, 08:09:32',
      id: 2,
      name: 'Running',
      updated_on: '06/26/2019, 08:09:32',
    },
  ],
};
