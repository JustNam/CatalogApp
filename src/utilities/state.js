export const initalCategoryState = {
  data: [],
};

export const initalItemState = {
  data: [],
  categoryId: 0,
  currentPage: 0,
  lastPage: 0,
};

export const initalUserState = {
  loggedIn: localStorage.getItem('accessToken') || false,
};