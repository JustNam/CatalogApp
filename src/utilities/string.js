export const convertCamelCaseToSnakeCase = string => (
  string.replace(/[A-Z]/g, match => `_${match.toLowerCase()}`)
);