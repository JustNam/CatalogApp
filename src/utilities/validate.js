export const validateItemTitle = (title) => {
  const { value } = title;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9\s]+$/)) {
    errors.push('Given title contains invalid character');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};

export const validateUsername = (username) => {
  const { value } = username;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9]+$/)) {
    errors.push('Given username contains invalid character');
  }
  if (!value.length > 30) {
    errors.push('Given username contains more than 30 characters');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};