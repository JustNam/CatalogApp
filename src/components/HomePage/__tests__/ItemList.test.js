import React from 'react';
import { shallow } from 'enzyme';
import ItemList from '../ItemList';
import { item, category, itemWithEmptyCategory } from '../../../utilities/sampleData';

describe('ItemList with full items', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      categoryId: 1,
      category,
      getItemsInCategoryWithPagination: jest.fn(() => Promise.resolve()),
      getCategories: jest.fn(() => Promise.resolve()),
      item,
    };
  };
  it('It should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  beforeEach(() => {
    setup();
    wrapper = shallow(<ItemList {...props} />);
    instance = wrapper.instance();
  });
  it('It should show a CreateItemModal', () => {
    wrapper.find('button[id="createButton"]').simulate('click');
    expect(wrapper.state().showInfoModal).toEqual(true);
  });
  it('It should close CreateItemModal when handleInfoClose is called', () => {
    instance.handleInfoClose();
    expect(wrapper.state().showInfoModal).toEqual(false);
  });
  it('It should open SuccessModal when handleSuccessShow is called', () => {
    instance.handleSuccessShow();
    expect(wrapper.state().showSuccessModal).toEqual(true);
  });
  it('It should close SuccessModal when handleSuccessClose is called', () => {
    instance.handleSuccessClose();
    expect(wrapper.state().showInfoModal).toEqual(false);
  });
  it('It should change page when one of buttons in pagination pane is clicked', () => {
    wrapper.find('PageItem[id=2]').simulate('click');
    expect(props.getItemsInCategoryWithPagination).toBeCalled();
  });
});


describe('ItemList with shortage items', () => {
  let wrapper;
  let props;
  const setup = () => {
    props = {
      categoryId: 2,
      category,
      getItemsInCategoryWithPagination: jest.fn(() => Promise.resolve()),
      getCategories: jest.fn(() => Promise.resolve()),
      item: itemWithEmptyCategory,
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<ItemList {...props} />);
  });

  it('It should show description which notifys that this category is empty', () => {
    expect(wrapper.find('a[className="preview-link"]').text()).toContain('This category is empty.');
  });
});
