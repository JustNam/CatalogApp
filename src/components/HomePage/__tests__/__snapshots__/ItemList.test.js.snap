// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList with full items It should render correctly 1`] = `ShallowWrapper {}`;
